staging/src/k8s.io/mount-utils/mount_linux.go

```diff
@@ -637,6 +640,57 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
 	return refs, nil
 }
 
+// IsMountPoint determines if a file is a mountpoint.
+// It first detects bind & any other mountpoints using
+// MountedFast function. If the MountedFast function returns
+// sure as true and err as nil, then a mountpoint is detected
+// successfully. When an error is returned by MountedFast, the
+// following is true:
+// 1. All errors are returned with IsMountPoint as false
+// except os.IsPermission.
+// 2. When os.IsPermission is returned by MountedFast, List()
+// is called to confirm if the given file is a mountpoint are not.
+//
+// os.ErrNotExist should always be returned a file does not exist
+// as callers have in past relied on this error and not fallback.
+//
+// When MountedFast returns sure as false and err as nil (eg: in
+// case of bindmounts on kernel version 5.10- ); mounter.List()
+// endpoint is called to enumerate all the mountpoints and check if
+// it is mountpoint match or not.
+func (mounter *Mounter) IsMountPoint(file string) (bool, error) {
+	isMnt, sure, isMntErr := mountinfo.MountedFast(file)
+	if sure == true && isMntErr == nil {
+		return isMnt, nil
+	}
+	if isMntErr != nil && errors.Is(isMntErr, fs.ErrPermission) {
+		// We were not allowed to do the simple stat() check, e.g. on NFS with
+		// root_squash. Fall back to /proc/mounts check below.
+		isMntErr = nil
+	}
+	if isMntErr != nil {
+		return false, isMntErr
+	}
+	// Resolve any symlinks in file, kernel would do the same and use the resolved path in /proc/mounts.
+	resolvedFile, err := filepath.EvalSymlinks(file)
+	if err != nil {
+		return false, err
+	}
+
+	// check all mountpoints since MountedFast is not sure.
+	// is not reliable for some mountpoint types.
+	mountPoints, mountPointsErr := mounter.List()
+	if mountPointsErr != nil {
+		return false, mountPointsErr
+	}
+	for _, mp := range mountPoints {
+		if isMountPointMatch(mp, resolvedFile) {
+			return true, nil
+		}
+	}
+	return false, nil
```

Comment:
Can we use `GetMounts` with a `SingleFileFilter` here, like `moby/sys/mountinfo` does?

The performance issue I'm trying to address in https://github.com/kubernetes/kubernetes/pull/109676 is caused by the repetitive reads of `/proc/mounts` at:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_linux.go#L544

And ultimately, optimizing the mount point checks is a better solution than trying to skip them (like #109676 tries to).

Reply:
We can't check that because unlike umount, we need to verify consistency in mountpoint check. GetMounts does not do a consistency check: https://github.com/moby/sys/blob/74ec3fe1325a016151d353bb6654d960042f9e96/mountinfo/mountinfo_linux.go#L19

because of that you will run into race condition with procfs; where a mountpoint may be incorrectly recognized. This is the main reason; we are trying to solve the problem in different ways.

Reply:
Apologies if this isn't the right place to ask, but I've crawled the git history and I'm still not understanding why the consistency check is necessary, particularly in the `CleanupMountPoint` code path.

Can you elaborate on the race condition(s)? Is there a race between `Mount` returning successfully and a respective entry in `/proc/mounts` existing? Or in the other direction; does a successful `Unmount` not result in the entry going away?

The consistency check seems like an intractable approach, because we're essentially requiring the system to become idle before we can perform further mounting operations.

Reply:
There was a bug in the kernel where /proc/mounts is read page by page. when 2 pages are read, then the 1st page can have an older snapshot and the 2nd page can be of an older snapshot reading to an inconsistent view of the mount-points. The assumption is that if 2 entire reads of /proc/mounts is the same and the mount entry is not present, then mountpoint is not present. This was solved in kernel 5.8+, commit https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=9f6c61f96f2d97cbb5f7fa85607bc398f843ff0f explains it much better.

Reply:
Thanks for the background!

It seems like we could break out of the consistency check if the mount was found in `/proc/mounts`, because this bug would only result in a mount being incorrectly omitted from `/proc/mounts`. I think such a change could be beneficial in the [first mount existence check in `CleanupMountPoint`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_helper_common.go#L90), because the happy path is when the `mountPath` *is* a mount point and is *not* a corrupted mount. I'll do some testing before drafting a PR -- and please let me know if this idea doesn't make sense 😅

Reply:
In the past, sig-storage has not accepted changes like removing a consistent read; although finding a case where consistent read is not required will be an awesome optimization.

This is the case that I was thinking can lead to issues.

Let's say we are reading /proc/mounts and it has 6 mount points initially.
A single page read can return 3 mountpoints. Between 2 page reads, the cursor might have changed 
and return a false positive that a mount-point does not exist when it actually does; leading
to data integrity issue.

As an example:


Suppose a page can have 3 mountpoints and we have 2 pages.

First content

SNP1, Page 1 :MP1, MP2, MP3
SNP1, Page 2  :MP4, MP5, MP6


2nd content (after MP1 has been unmounted which can happen between 2 page reads)

SNP2, Page 1 MP2, MP3, MP4 (1st page read can read MP4)
SNP2, Page 2 MP5, MP6 


the file content may have

MP1, MP2, MP3, MP5, MP6. In this case, MP4 is a mountpoint but has not been read.

Consistent read only partially solves the problem by comparing 2 consecutive reads but not wholly though 
as something like this might still happen.

Reply:
So, when determining if something is a mountpoint or not, if we encounter it at any point while parsing `/proc/mounts`, then we can return early -- we know that it must be a mount point. The bug can't result in mountpoints being mistakenly *included*, only mistakenly *excluded*.

For `CleanupMountPoint`, this only benefits the check before `Unmount`, but should be a meaningful optimization for the average case. After `Unmount`, however, we still need a consistent read to be confident that the path is *not* a mountpoint.

Looking at the code, this would be a significant refactor, because the consistent read operation is isolated from any notion of the target path we're checking. It's also not necessary if we have #109676 , because the consistent read is avoided both before and after `Unmount`. (Though some users could still benefit from it if their system does not have the desired "not mounted" behavior).

I'll draft a PR when I have a few minutes. I think it's worth doing, because the consistent read will be necessary for the foreseeable future, as the bug is only fixed in very recent kernels.

Reply:
I believe returning early works if you can make sure that you are not reading more than a single page.  AFAIK, strace can be really useful way to ensure that. If that is possible, code refactor is possible.